gnucash/gnome-utils/gnc-gtk-utils.c

```diff
@@ -332,3 +332,334 @@ gnc_get_dialog_widget_from_id (GtkDialog *dialog, const gchar *id)
+
+        if (g_strcmp0 (a_name, sm->search_action_name) == 0)
+        {
+            sm->menu_item = widget;
+            return;
+        }
+
+        if (accel_label)
+        {
+            // use gtk_label_get_text to get text with no underlines
+            const gchar *al_name = gtk_label_get_label (GTK_LABEL(accel_label));
+
+            if (g_strcmp0 (al_name, sm->search_action_name) == 0)
+            {
+                g_object_set_data_full (G_OBJECT(widget), "myaction-name",
+                                        g_strdup (sm->search_action_name), g_free);
+                sm->menu_item = widget;
+                return;
+            }
+        }
+
+        if (GTK_IS_CONTAINER(subMenu))
+            gtk_container_foreach (GTK_CONTAINER(subMenu),
+                                   search_menu_item_lookup, user_data);
+    }
+}
+
+
+static void
+search_menu_item_list (GtkWidget *widget, gpointer user_data)
+{
+    struct search_menu *sm = user_data;
+
+    if (GTK_IS_MENU_ITEM(widget))
+    {
+        GtkMenuItem* menuItem = GTK_MENU_ITEM(widget);
+        GtkWidget* subMenu = gtk_menu_item_get_submenu (menuItem);
+        const gchar *a_name = g_object_get_data (G_OBJECT(widget), "myaction-name");
+
+        sm->list = g_list_prepend (sm->list, widget);
+
+        if (!a_name)
+        {
+            GtkWidget *accel_label = gtk_bin_get_child (GTK_BIN(widget));
+
+            if (accel_label)
+            {
+                // use gtk_label_get_text to get text with no underlines
+                const gchar *al_name = gtk_label_get_label (GTK_LABEL(accel_label));
+
+                g_object_set_data_full (G_OBJECT(widget), "myaction-name",
+                                        g_strdup (al_name), g_free);
+            }
+        }
+
+        if (GTK_IS_CONTAINER(subMenu))
+            gtk_container_foreach (GTK_CONTAINER(subMenu),
+                                   search_menu_item_list, user_data);
+    }
+}
+
+/** Search the menu for menu item based on the label or action name
+ *
+ *  @param menu The menu widget.
+ *
+ *  @param action_name The GAction name.
+ *
+ *  @return The menu item widget or NULL.
+ */
+GtkWidget *
+gnc_find_menu_item (GtkWidget *menu, const gchar *action_name)
+{
+    struct search_menu sm;
+
+    g_return_val_if_fail (GTK_IS_WIDGET(menu), NULL);
+    g_return_val_if_fail (action_name != NULL, NULL);
+
+    sm.search_action_name = action_name;
+    sm.menu_item = NULL;
+
+    gtk_container_foreach (GTK_CONTAINER(menu), search_menu_item_lookup, &sm);
+
+    return sm.menu_item;
+}
+
+/** Return a list of menu items
+ *
+ *  @param menu The menu widget.
+ *
+ *  @return A GList of menu items or NULL.
+ */
+GList *
+gnc_menu_get_items (GtkWidget *menu)
+{
+    struct search_menu sm;
+
+    g_return_val_if_fail (GTK_IS_WIDGET(menu), NULL);
+
+    sm.search_action_name = NULL;
+    sm.menu_item = NULL;
+    sm.list = NULL;
+
+    gtk_container_foreach (GTK_CONTAINER(menu), search_menu_item_list, &sm);
```

Comment:
This use is part of `gnc_menu_get_items` which is returning a list of all entries so should be OK. I think your concern is with it's use in `gnc_find_menu_item` which calls `search_menu_item_lookup`. On line 415 I am testing for the search item being found and returning which seems to work as expected, am I missing some thing?

Reply:
Right, I put my comment in the wrong location.

What you are missing is that the function is still *called* for each and every menu_item, even after sm is set. It's true it will return early in the function, but for long menus it remains a big number of extra function calls. Again, I don't know for sure if it really would impact performance, but it may.

Reply:
There's `find_widget_func` in the same file in which the comparison function `(g_strcmp0 (name, id) == 0)` can be generalised.

https://github.com/Gnucash/gnucash/blob/7c29bf7efced3f817b0f5f1a8d0babf7c74d5de7/gnucash/gnome-utils/gnc-gtk-utils.c#L302

Reply:
Changed this to use a GList.